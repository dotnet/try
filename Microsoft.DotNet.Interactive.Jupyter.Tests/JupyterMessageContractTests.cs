// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using Assent;
using Microsoft.DotNet.Interactive.Jupyter.Protocol;
using Xunit;

namespace Microsoft.DotNet.Interactive.Jupyter.Tests
{
    public class JupyterMessageContractTests
    {
        private readonly Configuration _configuration;

        public JupyterMessageContractTests()
        {
            _configuration = new Configuration()
                .UsingExtension("json");

            _configuration = _configuration.SetInteractive(true);
        }

        [Fact]
        public void KernelInfoReply_contract_has_not_been_broken()
        {
            var socket = new TextSocket();
            var sender = new MessageSender(socket, new SignatureValidator("key", "HMACSHA256"));
            var kernelInfoReply = new KernelInfoReply(

                                       "5.3",
                                      ".NET",
                                      "0.0.3",
                                       new LanguageInfo(
                                           name: "C#",
                                           version: typeof(string).Assembly.ImageRuntimeVersion.Substring(1),
                                           mimeType: "text/x-csharp",
                                           fileExtension: ".cs",
                                           pygmentsLexer: "c#"
                                       ));
            var header = new Header(messageType: MessageTypeValues.KernelInfoReply, messageId: Guid.Empty.ToString(),
                version: "5.3", username: Constants.USERNAME, session: "test session",
                date: DateTime.MinValue.ToString("yyyy-MM-ddTHH:mm:ssZ"));
            var replyMessage = new Message(header, content: kernelInfoReply);
            sender.Send(replyMessage);

            var encoded = socket.GetEncodedMessage();
            this.Assent(encoded, _configuration);
        }

        [Fact]
        public void Execute_result_contract_has_not_been_broken()
        {
            var socket = new TextSocket();
            var sender = new MessageSender(socket, new SignatureValidator("key", "HMACSHA256"));
            var transient = new Dictionary<string, object> { { "display_id", "none" } };
            var output = "some result";
            var executeResult = new ExecuteResult(
                12,
                transient: transient,
                data: new Dictionary<string, object> {
                    { "text/html", output },
                    { "text/plain", output }

                });

            var header = new Header(messageType: MessageTypeValues.ExecuteResult, messageId: Guid.Empty.ToString(),
                version: "5.3", username: Constants.USERNAME, session: "test session",
                date: DateTime.MinValue.ToString("yyyy-MM-ddTHH:mm:ssZ"));

            var replyMessage = new Message(header, content: executeResult);

            sender.Send(replyMessage);

            var encoded = socket.GetEncodedMessage();
            this.Assent(encoded, _configuration);
        }

        [Fact]
        public void Display_data_contract_has_not_been_broken()
        {
            var socket = new TextSocket();
            var sender = new MessageSender(socket, new SignatureValidator("key", "HMACSHA256"));
            var transient = new Dictionary<string, object> { { "display_id", "none" } };
            var output = "some result";
            var displayData = new DisplayData(
                data: new Dictionary<string, object>
                {
                    {"text/html", output},
                    {"text/plain", output}
                },
                transient: transient);


            var header = new Header(messageType: MessageTypeValues.DisplayData, messageId: Guid.Empty.ToString(),
                version: "5.3", username: Constants.USERNAME, session: "test session",
                date: DateTime.MinValue.ToString("yyyy-MM-ddTHH:mm:ssZ"));
            var replyMessage = new Message(header, content: displayData);

            sender.Send(replyMessage);

            var encoded = socket.GetEncodedMessage();
            this.Assent(encoded, _configuration);
        }

        [Fact]
        public void Complete_reply_contract_has_not_been_broken()
        {
            var socket = new TextSocket();
            var sender = new MessageSender(socket, new SignatureValidator("key", "HMACSHA256"));

            var completeReply = new CompleteReply(0, 0, matches: new List<string> { "Write", "WriteLine" });

            var header = new Header(messageType: MessageTypeValues.CompleteReply, messageId: Guid.Empty.ToString(),
                version: "5.3", username: Constants.USERNAME, session: "test session",
                date: DateTime.MinValue.ToString("yyyy-MM-ddTHH:mm:ssZ"));

            var replyMessage = new Message(header, content: completeReply);
            sender.Send(replyMessage);

            var encoded = socket.GetEncodedMessage();
            this.Assent(encoded, _configuration);
        }

        [Fact]
        public void Update_data_contract_has_not_been_broken()
        {
            var socket = new TextSocket();
            var sender = new MessageSender(socket, new SignatureValidator("key", "HMACSHA256"));
            var transient = new Dictionary<string, object> { { "display_id", "none" } };
            var output = "some result";
            var displayData = new UpdateDisplayData
            (
                data: new Dictionary<string, object>
                {
                    { "text/html", output },
                    { "text/plain", output }
                },
                transient: transient
            );

            var header = new Header(messageType: MessageTypeValues.UpdateDisplayData, messageId: Guid.Empty.ToString(),
                version: "5.3", username: Constants.USERNAME, session: "test session",
                date: DateTime.MinValue.ToString("yyyy-MM-ddTHH:mm:ssZ"));

            var replyMessage = new Message(header, content: displayData);
            sender.Send(replyMessage);

            var encoded = socket.GetEncodedMessage();
            this.Assent(encoded, _configuration);
        }
    }
}