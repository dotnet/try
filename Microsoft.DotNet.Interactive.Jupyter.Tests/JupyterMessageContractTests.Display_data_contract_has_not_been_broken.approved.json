data: <IDS|MSG> more: True
data: c5121da74f5ffb064d676a281b3d48f643c9b26f60929966870dc52572cb7912 more: True
data: {"msg_id":"00000000-0000-0000-0000-000000000000","username":"dotnet_kernel","session":"test session","date":"0001-01-01T00:00:00Z","msg_type":"display_data","version":"5.3"} more: True
data: null more: True
data: {} more: True
data: {"data":{"text/html":"some result","text/plain":"some result"},"metadata":{},"transient":{"display_id":"none"}} more: False
